src/lib/krb5/ccache/cc_file.c

```diff
@@ -1002,12 +1018,144 @@ fcc_store(krb5_context context, krb5_ccache id, krb5_creds *creds)
     return set_errmsg_filename(context, ret ? ret : ret2, data->filename);
 }
 
-/* Non-functional stub for removing a cred from the cache file. */
+/*
+ * Overwrite cred in the ccache file with an entry that should not match any
+ * reasonable search.  Deletion is not guaranteed.  This algorithm originally
+ * from Heimdal.
+ */
+static krb5_error_code
+cred_delete(krb5_context context, krb5_ccache cache, krb5_cc_cursor *cursor,
+            krb5_creds *cred)
+{
+    krb5_error_code ret;
+    krb5_fcc_cursor *fcursor = *cursor;
+    fcc_data *fdata = cache->data;
+    struct k5buf expected = EMPTY_K5BUF, overwrite = EMPTY_K5BUF;
+    int fd = -1;
+    uint8_t *on_disk = NULL;
+    ssize_t rwret, start_offset;
+
+    k5_buf_init_dynamic_zap(&expected);
+    k5_buf_init_dynamic_zap(&overwrite);
+
+    /* This is what the cred looks like in the file (we hope). */
+
+    k5_marshal_cred(&expected, fcursor->version, cred);
+    ret = k5_buf_status(&expected);
+    if (ret)
+        goto cleanup;
+
+    /* Mark the cred expired so that it will hopefully be skipped over by any
+     * future match checks. */
+    cred->times.endtime = 0;
+
+    /* For config creds, endtimes aren't checked, so change the realm.
+     * Heimdal uses X-RMED-CONF here, so we do too. */
+    if (cred->server != NULL && cred->server->realm.length > 0 &&
+        strcmp(cred->server->realm.data, "X-CACHECONF:") == 0)
+        strcpy(cred->server->realm.data, "X-RMED-CONF:");
```

Comment:
I think any use of strcpy() will create a Coverity defect; you might need to use memcpy() instead.